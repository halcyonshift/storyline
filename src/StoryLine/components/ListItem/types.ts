import { ReactElement } from 'react'
import { ListItemProps as MuiListItemProps } from '@mui/material'

export interface ListItemProps extends MuiListItemProps {
    link?: string
    icon?: ReactElement
    primary: string
    secondary?: string
}
