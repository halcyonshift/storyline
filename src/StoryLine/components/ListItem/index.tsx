/** @format */

import { default as MuiListItem } from '@mui/material/ListItem'
import ListItemButton from '@mui/material/ListItemButton'
import ListItemIcon from '@mui/material/ListItemIcon'
import ListItemText from '@mui/material/ListItemText'
import { useTranslation } from 'react-i18next'
import { useNavigate } from 'react-router-dom'

import { ListItemProps } from './types'

const ListItem = ({ link, icon, text, divider }: ListItemProps) => {
    const navigate = useNavigate()
    const { t } = useTranslation()

    return (
        <MuiListItem disablePadding disableGutters divider={divider} onClick={() => navigate(link)}>
            <ListItemButton>
                <ListItemIcon>{icon}</ListItemIcon>
                <ListItemText primary={t(text)} />
            </ListItemButton>
        </MuiListItem>
    )
}

export default ListItem
