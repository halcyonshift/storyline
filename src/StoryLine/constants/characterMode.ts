export type CharacterModeType = 'PRIMARY' | 'SECONDARY' | 'TERTIARY'

export const CharacterMode = Object.freeze({
    PRIMARY: 'PRIMARY',
    SECONDARY: 'SECONDARY',
    TERTIARY: 'TERTIARY'
})
