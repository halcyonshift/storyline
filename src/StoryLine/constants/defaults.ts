export const DEFAULT_DISPLAY_MODE = 'light'
export const DEFAULT_FONT = 'roboto'
export const DEFAULT_FONT_SIZE = 14
export const DEFAULT_INDENT_PARAGRAPH = true
export const DEFAULT_LANGUAGE = 'en'
export const DEFAULT_LINE_SPACING = 1.5
export const DEFAULT_PARAGRAPH_SPACING = 2
export const DEFAULT_SPELL_CHECK = true
export const DEFAULT_AUTO_SAVE = true
export const DEFAULT_AUTO_BACKUP_FREQ = 0
export const DEFAULT_AUTO_BACKUP_MAX = 10
export const DEFAULT_PALETTE = 'indigo'
