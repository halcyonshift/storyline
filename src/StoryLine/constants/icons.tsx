import Icon from '@sl/components/Icon'

export const GLOBAL_ICONS = {
    add: <Icon name='Add' />,
    addImage: <Icon name='AddPhotoAlternate' />,
    back: <Icon name='ArrowBackIosNew' />,
    change: <Icon name='ChangeCircleOutlined' />,
    close: <Icon name='Close' />,
    delete: <Icon name='Delete' />,
    down: <Icon name='ArrowDownward' />,
    edit: <Icon name='EditOutlined' />,
    expand: <Icon name='ExpandMore' />,
    history: <Icon name='History' />,
    info: <Icon name='Info' />,
    next: <Icon name='ArrowForwardIos' />,
    open: <Icon name='FileOpen' />,
    random: <Icon name='Casino' />,
    up: <Icon name='ArrowUpward' />
}

export const CHARACTER_ICONS = {
    character: <Icon name='Group' />,
    addPrimary: <Icon name='PersonAddAlt1' />,
    addSecondary: <Icon name='PersonAddAlt' />,
    addTertiary: <Icon name='GroupAdd' />,
    primary: <Icon name='Person' />,
    secondary: <Icon name='PersonOutline' />,
    tertiary: <Icon name='PeopleAlt' />
}

export const CONNECTION_ICONS = {
    connections: <Icon name='Hub' />
}

export const IMPORTEXPORT_ICONS = {
    docx: <Icon name='Description' />,
    html: <Icon name='Html' />,
    importExport: <Icon name='ImportExport' />,
    import: <Icon name='ArrowDownward' />,
    pdf: <Icon name='PictureAsPdf' />
}

export const INSIGHT_ICONS = {
    insights: <Icon name='Insights' />
}

export const ITEM_ICONS = {
    item: <Icon name='Category' />,
    add: <Icon name='AddCircle' />
}

export const LOCATION_ICONS = {
    location: <Icon name='LocationOn' />,
    add: <Icon name='AddLocationAlt' />
}

export const NOTE_ICONS = {
    note: <Icon name='StickyNote2' />,
    add: <Icon name='PostAdd' />
}

export const OVERVIEW_ICONS = {
    summary: <Icon name='FormatListBulleted' />,
    filter: <Icon name='FilterList' />
}

export const RICHTEXT_ICONS = {
    bold: <Icon name='FormatBold' />,
    italic: <Icon name='FormatItalic' />,
    ul: <Icon name='FormatListBulleted' />,
    ol: <Icon name='FormatListNumbered' />,
    strike: <Icon name='FormatStrikethrough' />,
    underline: <Icon name='FormatUnderlined' />,
    alignLeft: <Icon name='FormatAlignLeft' />,
    alignRight: <Icon name='FormatAlignRight' />,
    alignCenter: <Icon name='FormatAlignCenter' />,
    alignJustify: <Icon name='FormatAlignJustify' />,
    tag: <Icon name='LocalOffer' />,
    version: <Icon name='RestorePage' />,
    redo: <Icon name='Redo' />,
    undo: <Icon name='Undo' />,
    save: <Icon name='Save' />,
    search: <Icon name='Search' />,
    excerpt: <Icon name='LabelImportant' />
}

export const SEARCH_ICONS = {
    search: <Icon name='Search' />,
    clear: <Icon name='Clear' />
}

export const SECTION_ICONS = {
    section: <Icon name='MenuBook' />,
    addPart: <Icon name='Queue' />,
    addChapter: <Icon name='LibraryBooks' />,
    addScene: <Icon name='PeopleAlt' />,
    version: <Icon name='RestorePage' />
}

export const SETTINGS_ICONS = {
    settings: <Icon name='Settings' />
}

export default {
    global: GLOBAL_ICONS,
    character: CHARACTER_ICONS,
    connection: CONNECTION_ICONS,
    importExport: IMPORTEXPORT_ICONS,
    insight: INSIGHT_ICONS,
    item: ITEM_ICONS,
    location: LOCATION_ICONS,
    note: NOTE_ICONS,
    overview: OVERVIEW_ICONS,
    search: SEARCH_ICONS,
    section: SECTION_ICONS,
    settings: SETTINGS_ICONS
}
