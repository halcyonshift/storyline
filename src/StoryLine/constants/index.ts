export const YYYYMMDD = 'yyyy-MM-dd'
