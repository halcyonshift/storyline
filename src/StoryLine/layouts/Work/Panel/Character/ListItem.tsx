import Box from '@mui/material/Box'
import { ListItem as MuiListItem } from '@mui/material'
import ListItemButton from '@mui/material/ListItemButton'
import ListItemText from '@mui/material/ListItemText'
import Stack from '@mui/material/Stack'
import Typography from '@mui/material/Typography'
import { useTranslation } from 'react-i18next'
import { GLOBAL_ICONS } from '@sl/constants/icons'
import CharacterModel from '@sl/db/models/CharacterModel'
import { status } from '@sl/theme/utils'
import TooltipIconButton from '@sl/components/TooltipIconButton'
import useTabs from '@sl/layouts/Work/useTabs'

const ListItem = ({ character }: { character: CharacterModel }) => {
    const tabs = useTabs()
    const { t } = useTranslation()

    return (
        <MuiListItem disablePadding disableGutters divider>
            <ListItemText
                primary={
                    <Box
                        className='flex justify-between align-middle'
                        sx={{ backgroundColor: status(character.status).color }}>
                        <ListItemButton
                            onClick={() =>
                                tabs.loadTab({
                                    id: character.id,
                                    label: character.displayName,
                                    link: `character/${character.id}`
                                })
                            }>
                            <Typography
                                variant='body1'
                                className='whitespace-nowrap text-ellipsis overflow-hidden'>
                                {character.displayName}
                            </Typography>
                        </ListItemButton>
                        <Stack spacing={0} direction='row'>
                            <TooltipIconButton
                                size='small'
                                text='layout.work.panel.character.edit'
                                link={`character/${character.id}/edit`}
                                icon={GLOBAL_ICONS.edit}
                            />
                            <TooltipIconButton
                                size='small'
                                text='layout.work.panel.character.delete'
                                icon={GLOBAL_ICONS.delete}
                                confirm={t('layout.work.panel.character.deleteConfirm', {
                                    name: character.displayName
                                })}
                                onClick={() => {
                                    tabs.removeTab(character.id)
                                    return character.delete()
                                }}
                            />
                        </Stack>
                    </Box>
                }
            />
        </MuiListItem>
    )
}

export default ListItem
