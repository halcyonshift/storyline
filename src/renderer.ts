import { init } from '@sentry/electron/renderer'
import { init as reactInit } from '@sentry/react'

if (parseInt(process.env.DEBUG) === 0) {
    init({ dsn: process.env.SENTRY }, reactInit)
}

import '@fontsource/roboto/300.css'
import '@fontsource/roboto/400.css'
import '@fontsource/roboto/500.css'
import '@fontsource/roboto/700.css'
import './index.css'
import './StoryLine'
