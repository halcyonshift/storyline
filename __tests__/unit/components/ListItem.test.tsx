import userEvent from '@testing-library/user-event'
import renderer from 'react-test-renderer'
import ListItem from '@sl/components/ListItem'
import { GLOBAL_ICONS } from '@sl/constants/icons'
import { render, screen } from '../test-utils'

const mockedUsedNavigate = jest.fn()

jest.mock('react-router-dom', () => ({
    ...(jest.requireActual('react-router-dom') as any),
    useNavigate: () => mockedUsedNavigate
}))

describe('<ListItem />', () => {
    it('Displays correct text', () => {
        render(<ListItem link='/' primary='this is text' />)
        expect(screen.getByText('this is text')).toBeTruthy()
    })

    it('Shows icon', () => {
        render(<ListItem link='/' primary='this is text' icon={GLOBAL_ICONS.add} />)
        expect(screen.getByLabelText('Add')).toBeTruthy()
    })

    it('Links as expected', async () => {
        render(<ListItem link='/a-link' primary='this is text' icon={GLOBAL_ICONS.add} />)
        await userEvent.click(screen.getByRole('button'))
        expect(mockedUsedNavigate).toHaveBeenCalledWith('/a-link')
    })

    it('renders correctly', () => {
        const tree = renderer.create(<ListItem link='/' primary='this is text' />).toJSON()
        expect(tree).toMatchSnapshot()
    })
})
