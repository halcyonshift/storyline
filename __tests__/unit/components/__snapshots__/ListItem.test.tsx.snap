// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListItem /> renders correctly 1`] = `
<li
  className="MuiListItem-root css-1nxmd3h-MuiListItem-root"
  disabled={false}
>
  <div
    className="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="MuiListItemText-root css-tlelie-MuiListItemText-root"
    >
      <span
        className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
      >
        this is text
      </span>
    </div>
  </div>
</li>
`;
